rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,237 @@
+  oak.handle.Sender sender = 2;
+}
+
+Message ReceiverChannelInfo {
+  // The name of the channel for the handle.
+  string channel_name = 1,
+
+  // The Receiver wrapping the read-half of the channel.
+  // Question: How do we define message_type?
+  oak.handle.Receiver receiver = 2;
+}
+
+Message NodeInitialisation {
+  // The Senders that the node expects.
+  repeated SenderChannelInfo = 1,
+
+  // The Receivers that the node expects.
+  repeated ReceiverChannelInfo = 2,
+
+  // A filtered map of config entries intended for the node.
+  ConfigMap config = 3
+}
+
+```
+
+### Accessing Channels
+
+All the channels needed by a node will be passed as part of the initialisation
+message. The channels are represented as repeating of `SenderChannelInfo`s and
+`ReceiverChannelInfo`s. The keys for each will be the name of the channel. This
+means that channels would need unique names in the manifest file. The downside
+of this approach is that the code needs to be aware of the names of the channels
+in the manifest file.
+
+This is not ideal, but we could work around it by adding a mapping as a config
+entry between internal names that the code understand and the channels names in
+the manifest file. We can then use this to support modification of the manifest
+file without having to recompile the WASM nodes.
+
+## Planning
+
+The initial implementation will be done as a `planner` example. It will create
+multiple test nodes and channels and connect these together. The example will
+also expose a gRPC method to the client so the client can request the manifest
+file. The client will compare the manifest file to the introspection data to
+test that the nodes were created as expected.
+
+Once the intial implementation is done, some of the common reusable code will be
+moved to the SDK. The fist example of such reusable code is support for node to
+accept an initial proto that contains a list of `Sender`s and `Receiver`s that
+wrap the handles for channels that should be available to the node, and
+potentially additional configuration files.
+
+The next step would be to update the existing examples to reuse the
+configuration-based `planner` node.
+
+## Future Work
+
+### Static analysis tools
+
+We can support application developers in finding potential issues in their
+applications through static analysis tools that can help detect common problems:
+
+- Check that all node and channel names are unique
+- Check that there is only one gRPC invocation receiver for each label
+- Check that there is only one HTTP invocation receiver for each label
+- Check that all nodes can write to handles they receive
+- Check that all nodes can read from handles they receive
+- Hihglight points where data is declassified to `Public`
+- Find nodes that can never receive data (unneeded nodes)
+- Find nodes where data will get stuck (dead ends)
+- Find risky patterns (e.g. diamond shaped graph)
+
+### Dynamic sub-graph creation
```

Comment:
Up to this point my main concern was how the planner would handle dynamic nodes, particularly with per-user labels.  Would it really be necessary to specify all possible users of an application up front in a node config file?

I *think* this subsection answers that concern, so it would be good to expand it and make some reference to it earlier in the doc&hellip;

Reply:
I was also confused up to this point as to how it would work (see my previous comments, which I will leave even though they are partially addressed here). Maybe mention this at the beginning, so that the reader is prepared? :)

Reply:
I will add a detailed design for per-label dynamic node creation.